### ISSUE TITLE
It's not possible to move Places root folders.

### ISSUE THREAD

**Issue by kayceegb** [Offerel/SyncMarks-Extension#28]
Hi,

I have this new one. 

I start 2 browser profiles, each with addon configured to use the php mode. 

Starting from, as always, an empty bookmark tree exported to webapp. 

I create a folder on one browser, which is well pushed to webapp, then I reload addon on the other browser to get the changes and nothing happens. No folder created. 

In debugger console, there is this
```
move is undefined background.js:892
    addPHPMarks moz-extension://6d62d872-dcfd-412c-9046-4df252796b2a/scripts/background.js:892
    apply self-hosted:1870
    applySafeWithoutClone resource://gre/modules/ExtensionCommon.jsm:614
    wrapPromise resource://gre/modules/ExtensionCommon.jsm:843
    withLastError resource://gre/modules/ExtensionCommon.jsm:762
    wrapPromise resource://gre/modules/ExtensionCommon.jsm:831
Unchecked lastError value: Error: It's not possible to move Places root folders. background.js:891
    addPHPMarks moz-extension://6d62d872-dcfd-412c-9046-4df252796b2a/scripts/background.js:891
```

You said you have a working configuration on your side. With all the changes already made, I wonder what is your setup because each time I try something "basic" that is not really working. But that's better than before for sure. 

K.

**Comment by Offerel** [comment]
There must be something messed with your bookmarks. Its not possible to move or edit systemfolders, like the named root folder. To identify whats going on in detail, i need more information. Are the some errors in webapp logfile? WHat do you have for entrys in your database and the logfile from the addon.

**Comment by kayceegb** [comment]
Well, I don't see how I could have something messed, as, I have done before, I always try without any bookmarks as a starting point (Remove and then export and then test). 

I don't remember having noted anything special except in XHR debug. So here we go verbosely:

Db after export with both clients synced:
```
sqlite> select * from bookmarks ORDER BY bmAdded;
bmID            bmParentID    bmIndex   bmTitle               bmType            bmURL                     bmAdded           bmModified        userID    bmAction
--------------  ------------  --------  --------------------  ----------------  ------------------------  ----------------  ----------------  --------  --------
menu________    root________  0         Bookmarks Menu        folder                                      1606994203249     1607091555593     1                 
toolbar_____    root________  1         Bookmarks Toolbar     folder                                      1606994203249     1607008750589     1                 
unfiled_____    root________  3         Other Bookmarks       folder                                      1606994203249     1607091554150     1                 
mobile______    root________  4         Mobile Bookmarks      folder                                      1606994203499     1607007502411     1                 
changes:   1   total_changes: 15
```

Client 1 creates folder
```
==> /var/log/bookmark.log <==
[04-Dec-2020 14:24:30] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin@shared.ff'
[04-Dec-2020 14:24:30] - [notice] - admin@shared.ff - 127.0.0.1 - Save session to database.
[04-Dec-2020 14:24:30] - [debug] - admin@shared.ff - 127.0.0.1 - UPDATE `users` SET `userLastLogin`= 1607091870, `sessionID` = '58td1lknb6uegojbfd9arg455h', `userOldLogin`= '1607085201' WHERE `userID` = '1'
[04-Dec-2020 14:24:30] - [notice] - admin@shared.ff - 127.0.0.1 - Get userdata from the database
[04-Dec-2020 14:24:30] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin@shared.ff'
[04-Dec-2020 14:24:30] - [notice] - admin@shared.ff - 127.0.0.1 - Try to add folder 'New Folder'
[04-Dec-2020 14:24:30] - [debug] - admin@shared.ff - 127.0.0.1 - {"id":"MEdO_ED6m-1F","title":"New Folder","type":"folder","folder":"unfiled_____","nfolder":"Other Bookmarks","added":1607091869980}
[04-Dec-2020 14:24:30] - [notice] - admin@shared.ff - 127.0.0.1 - Try to find if this folder exists already
[04-Dec-2020 14:24:30] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT COUNT(*) AS bmCount, bmAction, bmID  FROM `bookmarks` WHERE `bmTitle` = 'New Folder' AND `bmParentID` = 'unfiled_____' AND `userID` = 1;
[04-Dec-2020 14:24:30] - [notice] - admin@shared.ff - 127.0.0.1 - Undefined variable: fdExist in /var/www/html/bmsync/index.php on line 918
[04-Dec-2020 14:24:30] - [notice] - admin@shared.ff - 127.0.0.1 - Get folder data for adding folder
[04-Dec-2020 14:24:30] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT IFNULL(MAX(`bmIndex`),-1) + 1 AS `nindex`, `bmParentId` FROM `bookmarks` WHERE `bmParentId` = 'unfiled_____' AND `userID` = 1;
[04-Dec-2020 14:24:30] - [debug] - admin@shared.ff - 127.0.0.1 - INSERT INTO `bookmarks` (`bmID`,`bmParentID`,`bmIndex`,`bmTitle`,`bmType`,`bmAdded`,`userID`) VALUES ('MEdO_ED6m-1F', 'unfiled_____', 0, 'New Folder', 'folder', 1607091869980, 1)

==> /var/log/nginx/access.log <==
127.0.0.1 - admin@shared.ff [04/Dec/2020:15:24:30 +0100] "POST /bmsync/index.php HTTP/1.1" 200 11 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0"

==> /var/log/bookmark.log <==
[04-Dec-2020 14:24:33] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin@shared.ff'
[04-Dec-2020 14:24:33] - [notice] - admin@shared.ff - 127.0.0.1 - Get userdata from the database
[04-Dec-2020 14:24:33] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin@shared.ff'
[04-Dec-2020 14:24:33] - [notice] - admin@shared.ff - 127.0.0.1 - Edit folder request, try to find the folder...
[04-Dec-2020 14:24:33] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT * FROM `bookmarks` WHERE `bmIndex` >= 0 AND `bmType` = 'folder' AND `bmParentID` = 'unfiled_____' AND `userID` = 1;
[04-Dec-2020 14:24:33] - [notice] - admin@shared.ff - 127.0.0.1 - Unique folder found, edit the folder
[04-Dec-2020 14:24:33] - [debug] - admin@shared.ff - 127.0.0.1 - UPDATE `bookmarks` SET `bmAction` = NULL, `bmTitle` = 'f1' WHERE `bmID` = 'MEdO_ED6m-1F' AND userID = 1;

==> /var/log/nginx/access.log <==
127.0.0.1 - admin@shared.ff [04/Dec/2020:15:24:33 +0100] "POST /bmsync/index.php HTTP/1.1" 200 5 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0"
```

Db has:
```
sqlite> select * from bookmarks ORDER BY bmAdded;
bmID            bmParentID    bmIndex   bmTitle               bmType            bmURL                     bmAdded           bmModified        userID    bmAction
--------------  ------------  --------  --------------------  ----------------  ------------------------  ----------------  ----------------  --------  --------
menu________    root________  0         Bookmarks Menu        folder                                      1606994203249     1607091555593     1                 
toolbar_____    root________  1         Bookmarks Toolbar     folder                                      1606994203249     1607008750589     1                 
unfiled_____    root________  3         Other Bookmarks       folder                                      1606994203249     1607091554150     1                 
mobile______    root________  4         Mobile Bookmarks      folder                                      1606994203499     1607007502411     1                 
MEdO_ED6m-1F    unfiled_____  0         f1                    folder                                      1607091869980                       1                 
changes:   1   total_changes: 15
```

Client 2 reload addon, folder is not there

webapp log:
```
==> /var/log/bookmark.log <==
[04-Dec-2020 14:27:24] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin@shared.ff'
[04-Dec-2020 14:27:24] - [notice] - admin@shared.ff - 127.0.0.1 - Save session to database.
[04-Dec-2020 14:27:24] - [debug] - admin@shared.ff - 127.0.0.1 - UPDATE `users` SET `userLastLogin`= 1607092044, `sessionID` = 'hr56ap7tjb49s1h7iqn0s9aam2', `userOldLogin`= '1607091870' WHERE `userID` = '1'
[04-Dec-2020 14:27:24] - [notice] - admin@shared.ff - 127.0.0.1 - Get userdata from the database
[04-Dec-2020 14:27:24] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin@shared.ff'
[04-Dec-2020 14:27:25] - [notice] - admin@shared.ff - 127.0.0.1 - Browser startup sync started, get client data
[04-Dec-2020 14:27:25] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT `lastseen` FROM `clients` WHERE `cid` = 'b35f6092-21b2-47f0-8abc-eb2c17f48729' AND `uid` = 1 AND `ctype` = 'Firefox'
[04-Dec-2020 14:27:25] - [notice] - admin@shared.ff - 127.0.0.1 - Get changed bookmarks for client b35f6092-21b2-47f0-8abc-eb2c17f48729
[04-Dec-2020 14:27:25] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT b.bmID AS fdID, b.bmTitle AS fdName, b.bmIndex AS fdIndex, a.bmIndex, a.bmTitle, a.bmType, a.bmURL, a.bmAdded, a.bmModified, a.bmAction FROM bookmarks a INNER JOIN bookmarks b ON b.bmID = a.bmParentID WHERE (a.bmAdded >= 1607091563992 AND a.userID = 1) OR (a.bmAction = 1 AND a.bmAdded >= 1607091563992 AND a.userID = 1)
[04-Dec-2020 14:27:25] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT * FROM `clients` WHERE `cid` = 'b35f6092-21b2-47f0-8abc-eb2c17f48729' AND uid = 1
[04-Dec-2020 14:27:25] - [notice] - admin@shared.ff - 127.0.0.1 - Updating lastlogin for client b35f6092-21b2-47f0-8abc-eb2c17f48729.
[04-Dec-2020 14:27:25] - [notice] - admin@shared.ff - 127.0.0.1 - Try to find bookmarks, which could be completely deleted
[04-Dec-2020 14:27:25] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT bmID FROM bookmarks WHERE bmAdded <= (SELECT MIN(lastseen) FROM clients WHERE uid = 1 AND lastseen > 1) AND bmAction = 1
[04-Dec-2020 14:27:25] - [notice] - admin@shared.ff - 127.0.0.1 - No bookmarks found to delete from the database
[04-Dec-2020 14:27:25] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin@shared.ff'

==> /var/log/nginx/access.log <==
127.0.0.1 - admin@shared.ff [04/Dec/2020:15:27:25 +0100] "POST /bmsync/index.php?t=0.9225574801892117 HTTP/1.1" 200 197 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:78.0) Gecko/20100101 Firefox/78.0"

==> /var/log/bookmark.log <==
[04-Dec-2020 14:27:25] - [notice] - admin@shared.ff - 127.0.0.1 - Get userdata from the database
[04-Dec-2020 14:27:25] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin@shared.ff'
[04-Dec-2020 14:27:25] - [notice] - admin@shared.ff - 127.0.0.1 - Try to get list of clients.
[04-Dec-2020 14:27:25] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT cid, IFNULL(cname, cid) cname, ctype, lastseen FROM clients WHERE uid = 1 AND NOT cid = 'b35f6092-21b2-47f0-8abc-eb2c17f48729' ORDER BY 2 COLLATE NOCASE ASC;
[04-Dec-2020 14:27:25] - [notice] - admin@shared.ff - 127.0.0.1 - Found 1 clients. Send list to requesting client.
[04-Dec-2020 14:27:25] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin@shared.ff'

==> /var/log/nginx/access.log <==
127.0.0.1 - admin@shared.ff [04/Dec/2020:15:27:25 +0100] "POST /bmsync/index.php HTTP/1.1" 200 144 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:78.0) Gecko/20100101 Firefox/78.0"

==> /var/log/bookmark.log <==
[04-Dec-2020 14:27:25] - [notice] - admin@shared.ff - 127.0.0.1 - Get userdata from the database
[04-Dec-2020 14:27:25] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin@shared.ff'
[04-Dec-2020 14:27:25] - [notice] - admin@shared.ff - 127.0.0.1 - Get pushed site from clients.
[04-Dec-2020 14:27:25] - [debug] - admin@shared.ff - 127.0.0.1 - SELECT * FROM `notifications` WHERE `nloop` = 1 AND `userID` = 1 AND `repeat` IN ('b35f6092-21b2-47f0-8abc-eb2c17f48729','0');
[04-Dec-2020 14:27:25] - [notice] - admin@shared.ff - 127.0.0.1 - Found 0 links. Will push them to the client.
```

Log from addon:
```
12/4/2020, 3:27:24 PM - Info: Initiate PHP startup sync
12/4/2020, 3:27:24 PM - Info: Get list of clients.
12/4/2020, 3:27:24 PM - Info: Get notifications for current client.
12/4/2020, 3:27:24 PM - Info: Initiate startup sync
12/4/2020, 3:27:25 PM - 1 changes on the server since the last sync.
12/4/2020, 3:27:25 PM - Info: Changed bookmark null is in systemfolder
12/4/2020, 3:27:25 PM - Info: Try to add bookmark null
12/4/2020, 3:27:25 PM - Info: List of 2 clients retrieved successfully.
```

And XHR debug:
Request:
```
client | "b35f6092-21b2-47f0-8abc-eb2c17f48729"
-- | --
caction | "startup"
```
Response:
```
    0	Object { fdID: "unfiled_____", fdName: "Other Bookmarks", fdIndex: "3", … }
    fdID	"unfiled_____"
    fdName	"Other Bookmarks"
    fdIndex	"3"
    bmIndex	"0"
    bmTitle	"f1"
    bmType	"folder"
    bmURL	null
    bmAdded	"1607091869980"
    bmModified	null
    bmAction	null
```
```
move is undefined background.js:892
    addPHPMarks moz-extension://6d62d872-dcfd-412c-9046-4df252796b2a/scripts/background.js:892
    apply self-hosted:1870
    applySafeWithoutClone resource://gre/modules/ExtensionCommon.jsm:614
    wrapPromise resource://gre/modules/ExtensionCommon.jsm:843
    withLastError resource://gre/modules/ExtensionCommon.jsm:762
    wrapPromise resource://gre/modules/ExtensionCommon.jsm:831
```
```
Unchecked lastError value: Error: It's not possible to move Places root folders. background.js:891
    addPHPMarks moz-extension://6d62d872-dcfd-412c-9046-4df252796b2a/scripts/background.js:891
```

**Comment by Offerel** [comment]
For me, the DB and the response to the second client looks fine. There should be no move trigger for a systemfolder. But there is wrong identified folder according to the log from the second addon. This should not lead to problems, but i have changed now, how i detect folder and bookmark better. I will include this in the next version. For the meantime you can ignore this.

**Comment by kayceegb** [comment]
I agree. The create and "send modifications" processes seems right. The informations received by addon are correct. It's just it is doing something wrong with valid information. 

You say it should not lead to problems, but actually it does. The sync process like it is now is broken. You create a folder and a bookmark on one side and it is not replicated on the other side. I don't know how I can ignore this.

**Comment by kayceegb** [comment]
I seen your last commit and had time to test. 

Here is what it reports. 

```
Uncaught TypeError: bookmark.bmURL is null
    addPHPMarks moz-extension://661aea37-1ee8-4551-950e-f329229ea718/scripts/background.js:885
    addPHPMarks moz-extension://661aea37-1ee8-4551-950e-f329229ea718/scripts/background.js:834
    onload moz-extension://661aea37-1ee8-4551-950e-f329229ea718/scripts/background.js:700
background.js:885:8
```

And folder is not synced once client 2 starts.

**Comment by kayceegb** [comment]
I have made some progress on my side. I have done this almost by mimetism. Folders on client 2 are present after restart unfortunatly in the wrong order(index) for the moment (I have to see how to trick the async mode).

```
diff --git a/scripts/background.js b/scripts/background.js
index e217eca..a9e9834 100644
--- a/scripts/background.js
+++ b/scripts/background.js
@@ -881,22 +881,43 @@ function addPHPMarks(bArray) {
                                });
                        }
                        else {
-                               loglines = logit('Info: Changed bookmark '+bookmark.bmURL+' is in systemfolder');
-                               if(bookmark.bmURL != '') {
-                                       loglines = logit('Info: Try to add bookmark '+bookmark.bmURL);
-                                       chrome.bookmarks.search({url: bookmark.bmURL}, function(bookmarkItems) {
+                               if(bookmark.bmType == "bookmark"){
+                                       loglines = logit('Info: Changed bookmark '+bookmark.bmURL+' is in systemfolder');
+                                       if(bookmark.bmURL != '') {
+                                               loglines = logit('Info: Try to add bookmark '+bookmark.bmURL);
+                                               chrome.bookmarks.search({url: bookmark.bmURL}, function(bookmarkItems) {
+                                                       if (bookmarkItems.length) {
+                                                               if(bookmarkItems[0].parentId != bookmark.fdID) {
+                                                                       chrome.bookmarks.onMoved.removeListener(onMovedCheck);
+                                                                       chrome.bookmarks.move(bookmarkItems[0].id, {parentId: bookmark.fdID}, function(move) {
+                                                                               loglines = logit('Info: '+move.url + " moved to folder " + bookmark.fdName);
+                                                                               chrome.bookmarks.onMoved.addListener(onMovedCheck);
+                                                                       });
+                                                               }
+                                                       } else {
+                                                               chrome.bookmarks.onCreated.removeListener(onCreatedCheck);
+                                                               chrome.bookmarks.create({type: bookmark.bmType, parentId: bookmark.fdID, title: bookmark.bmTitle, url: bookmark.bmURL}, function() {
+                                                                       loglines = logit('Info: '+bookmark.bmURL + " added as new bookmark.");
+                                                                       chrome.bookmarks.onCreated.addListener(onCreatedCheck);
+                                                               });
+                                                       }
+                                               });
+                                       }
+                               }
+                               else if(bookmark.bmType == "folder"){
+                                       chrome.bookmarks.search({title: bookmark.bmTitle}, function(bookmarkItems) {
                                                if (bookmarkItems.length) {
                                                        if(bookmarkItems[0].parentId != bookmark.fdID) {
                                                                chrome.bookmarks.onMoved.removeListener(onMovedCheck);
                                                                chrome.bookmarks.move(bookmarkItems[0].id, {parentId: bookmark.fdID}, function(move) {
-                                                                       loglines = logit('Info: '+move.url + " moved to folder " + bookmark.fdName);
+                                                                       loglines = logit('Info: folder '+move.title + " moved to folder " + bookmark.fdName);
                                                                        chrome.bookmarks.onMoved.addListener(onMovedCheck);
                                                                });
                                                        }
                                                } else {
                                                        chrome.bookmarks.onCreated.removeListener(onCreatedCheck);
-                                                       chrome.bookmarks.create({type: bookmark.bmType, parentId: bookmark.fdID, title: bookmark.bmTitle, url: bookmark.bmURL}, function() {
-                                                               loglines = logit('Info: '+bookmark.bmURL + " added as new bookmark.");
+                                                       chrome.bookmarks.create({type: bookmark.bmType, parentId: bookmark.fdID, title: bookmark.bmTitle}, function() {
+                                                               loglines = logit('Info: '+bookmark.bmTitle + " added as new folder.");
                                                                chrome.bookmarks.onCreated.addListener(onCreatedCheck);
                                                        });
                                                }
```

**Comment by kayceegb** [comment]
Still the same error after last commit

```
Uncaught TypeError: bookmark.bmURL is null
    addPHPMarks moz-extension://ddd54a51-f329-44d8-8469-e06193db7e9c/scripts/background.js:884
    addPHPMarks moz-extension://ddd54a51-f329-44d8-8469-e06193db7e9c/scripts/background.js:834
    onload moz-extension://ddd54a51-f329-44d8-8469-e06193db7e9c/scripts/background.js:700
```

**Comment by kayceegb** [comment]
Even with something like this it's not working
```
diff --git a/scripts/background.js b/scripts/background.js
index c02cee1..a109901 100644
--- a/scripts/background.js
+++ b/scripts/background.js
@@ -881,7 +881,7 @@ function addPHPMarks(bArray) {
                                });
                        }
                        else {
-                               if(bookmark.bmURL.length != null) {
+                               if(bookmark.bmURL != null) {
                                        loglines = logit('Info: Try to add bookmark '+bookmark.bmURL);
                                        chrome.bookmarks.search({url: bookmark.bmURL}, function(bookmarkItems) {
                                                if (bookmarkItems.length) {
```

The error is not there anymore but folder is not created and in log reported as 
> Info: Changed bookmark null in userfolder

**Comment by Offerel** [comment]
As i said before, i cant reproduce your issue. I fixed only the error message. That's all what i can do here. If you install the addon via AMO, it should work as expected. I know there are issue, when you install or test it via web-ext or sideload it or install it as a temporary addon. But since these are non standard ways of installing an addon, i cant do much against that.

**Comment by kayceegb** [comment]
I don't know how you tried to reproduce the issue as latest changes exclude folder creation processing in addon. You say you fixed the error message but at least here the error message is still there if you want to sync a folder, I repeat a folder and not a bookmark. 

What is your setup for testing ?

**Comment by Offerel** [comment]
What do you mean with setup for testing? The addon is installed from AMO, as every user does this. Inside the addon, i enabled PHP and all options. Nothing special there.

**Comment by kayceegb** [comment]
I wanted to know the environment with which you do testing. You said If I remember correctly you are on Windows. 10, Right ? Which revision ? And firefox version ? Maybe I can try to see where is the difference. 
You say nothing special, but do you clear the bookmarks in receiving side before test (can be done with the addon's button "remove") ? 

In addition, I reverted to master rev to try to debug the move operation. The changes are:

```diff
diff --git a/scripts/background.js b/scripts/background.js
index e217eca..991a2ff 100644
--- a/scripts/background.js
+++ b/scripts/background.js
@@ -881,12 +881,16 @@ function addPHPMarks(bArray) {
                                });
                        }
                        else {
-                               loglines = logit('Info: Changed bookmark '+bookmark.bmURL+' is in systemfolder');
+                               loglines = logit('Info: Changed DBG bookmark '+bookmark.bmTitle+' is in systemfolder');
                                if(bookmark.bmURL != '') {
-                                       loglines = logit('Info: Try to add bookmark '+bookmark.bmURL);
+                                       loglines = logit('Info: Try to add bookmark '+bookmark.bmTitle);
                                        chrome.bookmarks.search({url: bookmark.bmURL}, function(bookmarkItems) {
                                                if (bookmarkItems.length) {
+                                                       for(i=0; i<bookmarkItems.length; i++){
+                                                               loglines = logit('DBG: '+i+' : '+bookmarkItems[i].title);
+                                                       }
                                                        if(bookmarkItems[0].parentId != bookmark.fdID) {
+                                                               loglines = logit('DBG: '+bookmarkItems[0].parentId+' - '+bookmark.fdID);
                                                                chrome.bookmarks.onMoved.removeListener(onMovedCheck);
                                                                chrome.bookmarks.move(bookmarkItems[0].id, {parentId: bookmark.fdID}, function(move) {
                                                                        loglines = logit('Info: '+move.url + " moved to folder " + bookmark.fdName);
```

The change received from backend is still the same:
```
    0	Object { fdID: "unfiled_____", fdName: "Other Bookmarks", fdIndex: "3", … }
    fdID	"unfiled_____"
    fdName	"Other Bookmarks"
    fdIndex	"3"
    bmIndex	"0"
    bmTitle	"f1"
    bmType	"folder"
    bmURL	null
    bmAdded	"1607280740719"
    bmModified	null
    bmAction	null
```

And addon's log have this:
```
12/6/2020, 8:30:23 PM - Info: Initiate PHP startup sync
12/6/2020, 8:30:23 PM - Info: Get list of clients.
12/6/2020, 8:30:23 PM - Info: Get notifications for current client.
12/6/2020, 8:30:23 PM - Info: Initiate startup sync
12/6/2020, 8:30:23 PM - 1 changes on the server since the last sync.
12/6/2020, 8:30:23 PM - Info: Changed DBG bookmark f1 is in systemfolder
12/6/2020, 8:30:23 PM - Info: Try to add bookmark f1
12/6/2020, 8:30:23 PM - DBG: 0 : Bookmarks Menu
12/6/2020, 8:30:23 PM - DBG: 1 : Bookmarks Toolbar
12/6/2020, 8:30:23 PM - DBG: 2 : tags
12/6/2020, 8:30:23 PM - DBG: 3 : Other Bookmarks
12/6/2020, 8:30:23 PM - DBG: 4 : Mobile Bookmarks
12/6/2020, 8:30:23 PM - DBG: root________ - unfiled_____
```

So, we can see that the try to do the move operation, move which is not allowed, corresponds to what code says to do. It maybe not happens always, but it seems I am able to trigger this. 

The following change, mitigates the problem:
```diff
diff --git a/scripts/background.js b/scripts/background.js
index e217eca..5fb1491 100644
--- a/scripts/background.js
+++ b/scripts/background.js
@@ -884,7 +884,7 @@ function addPHPMarks(bArray) {
                                loglines = logit('Info: Changed bookmark '+bookmark.bmURL+' is in systemfolder');
                                if(bookmark.bmURL != '') {
                                        loglines = logit('Info: Try to add bookmark '+bookmark.bmURL);
-                                       chrome.bookmarks.search({url: bookmark.bmURL}, function(bookmarkItems) {
+                                       chrome.bookmarks.search({url: bookmark.bmURL, title: bookmark.bmTitle}, function(bookmarkItems) {
                                                if (bookmarkItems.length) {
                                                        if(bookmarkItems[0].parentId != bookmark.fdID) {
                                                                chrome.bookmarks.onMoved.removeListener(onMovedCheck);
```

And the folder is created on receiving side. 

Another approach would be to differentiate between bookmark processing and folder processing like the changes I tried at post https://github.com/Offerel/SyncMarks-Extension/issues/28#issuecomment-739413608

I don't know if it more clear for you.

**Comment by Offerel** [comment]
I use a handful different computers. Most have some revision of Windows 10. Firefox is mostly the latest stable release, sometimes nightly. Sometimes Chrome or Edge. I have never seen a problem like this. Removing bookmarks is the default, before receiving a new bookmark set.

**Comment by kayceegb** [comment]
So, I have a test environment with latest w10 and latest firefox. 

Test procedure:

client 1:
install addon from amo
in addon, configure and save to validate credentials
in addon, click remove to remove all bookmarks
in addon, click export to export to backend
in firefox, create a bookmark
in firefox create a folder

check webapp to confirm everything is ok

client 2:
install addon from amo
in addon, click remove to remove all bookmarks
in addon, configure and save to validate credentials
close firefox
open firefox again

bookmark is synced
folder is not synced

and in debug there is 
```
Uncaught TypeError: bookmark.bmURL is null
    addPHPMarks moz-extension://00e54b59-5f43-4e9e-bcef-d23b19bb78f4/scripts/background.js:885
    addPHPMarks moz-extension://00e54b59-5f43-4e9e-bcef-d23b19bb78f4/scripts/background.js:834
    onload moz-extension://00e54b59-5f43-4e9e-bcef-d23b19bb78f4/scripts/background.js:700
background.js:885:8
```

**Comment by Offerel** [comment]
It looks like you have installed a version with the old folder detection code. But as i noted before, this isn't the problem and you can ignore this. I will look in AMO, if i can upload or free the latest version. Maybe it's not reviewed yet.

Anyway instead of this, i have no idea whats going with your bookmarks. Something which i cant identify from here prevents creation of bookmarks/folders. The only and last idea form my side is the following: I have added a function to save a file export_<username>_<timestamp>.json when loglevel is set to 9. This saves the JSON, which is send to the browser in that file. If you send me that file or add it here or at pastebin, i can check this file on my side. Besides this, i cant do anything for you, I'm out of ideas.

If generating this file is useful, i will look for a way to enable/disable creation of this file more comfortable.

**Comment by kayceegb** [comment]
When I tested before this day, the version available from AMO was 1.17.6.3. And the error was Uncaught TypeError: bookmark.bmURL is null.

This is the response received by addon. That corresponds to the test done earlier this day
```
    0	Object { fdID: "unfiled_____", fdName: "Other Bookmarks", fdIndex: "3", … }
    fdID	"unfiled_____"
    fdName	"Other Bookmarks"
    fdIndex	"3"
    bmIndex	"0"
    bmTitle	"Issues · Offerel/SyncMarks-Webapp · GitHub"
    bmType	"bookmark"
    bmURL	"https://github.com/Offerel/SyncMarks-Webapp/issues?q=is:issuesort:updated-descis:open"
    bmAdded	"1607377291377"
    bmModified	null
    bmAction	null
    1	Object { fdID: "unfiled_____", fdName: "Other Bookmarks", fdIndex: "3", … }
    fdID	"unfiled_____"
    fdName	"Other Bookmarks"
    fdIndex	"3"
    bmIndex	"1"
    bmTitle	"f1"
    bmType	"folder"
    bmURL	null
    bmAdded	"1607377308692"
    bmModified	null
    bmAction	null
```

As you can see, there is data for one bookmark and for one folder. 

The code responsible for the error shown was this one
> if(bookmark.bmURL.length != null) {

From data below, you can see that bmURL for a folder is null so the code generated an error because length function can't return a length for a null string and that was reported by addon debugger and that prevented folder sync processing. 

Now with the last version, 1.17.6.4, you have changed the code to
```
-                               if(bookmark.bmURL.length != null) {
+                               if(bookmark.bmURL != null) {
```

So the code is correct now and does not throw an error anymore. 

But with this change and even before, I don't see how you can say that folder is processed correctly. It's just **impossible**. 

Once again, folder data has bmURL as `null` so it is excluded from processing because of this line
>                                if(bookmark.bmURL != null) {

I don't see anywhere code for folder processing. There is no **else** below line 884 and corresponding to line 884. 
```
883             else {
 884                 if(bookmark.bmURL != null) {
 885 +-- 18 lignes : loglines = logit('Info: Try to add bookmark '+bookmark.bmURL);---------------------------------------------------------------------------------------------------
 903                 }
 904             }
```

Or maybe I don't see it. So can you be gentle and tell me how is folder processing done and where is the code related to folder processing ? I don't see any.

**Comment by kayceegb** [comment]
Any feedback on this ? 

I would like to understand where I am wrong.
